main.cpp

```diff
@@ -435,13 +444,31 @@ void Mainloop::loop()
                 handle_tcp_connection();
                 continue;
             }
-            Endpoint *e = static_cast<Endpoint*>(events[i].data.ptr);
+            Pollable *p = static_cast<Pollable *>(events[i].data.ptr);
+
+            if (Endpoint *e = dynamic_cast<Endpoint *>(p)) {
+                if (events[i].events & EPOLLIN)
+                    handle_read(e);
+
+                if (events[i].events & EPOLLOUT)
+                    handle_canwrite(e);
+
+                continue;
+            }
+
+            if (Timeout *t = dynamic_cast<Timeout *>(p)) {
+                uint64_t val = 0;
```

Comment:
You don't use dynamic_cast here.... Just make Timeout implement the same interface with `handle_read()` and an empty `handle_canwrite()`.... this implementation you are doing here should be the `handle_read()` of the Timeout.